cas-server-webapp-validation/src/main/java/org/jasig/cas/web/AbstractServiceValidateController.java

```diff
@@ -128,13 +151,38 @@ protected Credential getServiceCredentialsFromRequest(final WebApplicationServic
                 verifyRegisteredServiceProperties(registeredService, service);
                 return new HttpBasedServiceCredential(new URL(pgtUrl), registeredService);
             } catch (final Exception e) {
-                logger.error("Error constructing pgtUrl", e);
+                logger.error("Error constructing {}", CasProtocolConstants.PARAMETER_PROXY_CALLBACK_URL, e);
             }
         }
 
         return null;
     }
 
+
+    /**
+     * Validate authentication context pair.
+     *
+     * @param assertion the assertion
+     * @param request   the request
+     * @return the pair
+     */
+    protected Pair<Boolean, Optional<MultifactorAuthenticationProvider>> validateAuthenticationContext(
+            final Assertion assertion, final HttpServletRequest request) {
+        final String requestedContext = request.getParameter(this.authenticationContextParameter);
```

Comment:
couldn't we also load the `requestedContext` based on the configuration within the `RegisteredService` itself?

This would allow most services that want to enforce MFA to only need to enable it centrally in the ServiceRegistry and not need to specifically request it when validating tickets.